### ISSUE TITLE
Patched PragmataPro glyphs are being rendered sliced

### ISSUE THREAD

**Issue by ToasterDEV** [ryanoasis/nerd-fonts#682]
#### 🎯 Subject of the issue

After trying to patch the PragmataPro family of fonts with the following command:
```
docker run -v /mnt/c/Users/ToasterDEV/OneDrive/Documents/Backups/Typography/PragmataPro:/in -v /mnt/c/Users/ToasterDEV/OneDrive/Documents/Backups/Typography/PragmataProNerdFonts/out:/out nerdfonts/patcher -l -w -c -s --progressbars --careful
```
The glyphs are successfully added, but are rendered halfway as shown in the attached screenshot, regardless of both the Windows compatibility or the single width flags being enabled while patching the font.
#### 🔧 Your Setup

- _Which font are you using:_ `PragmataProMonoLiga NF.ttf`
- _Which terminal emulator are you using:_ Windows Terminal 1.11.2911.0
- _Are you using OS X, Linux or Windows?_ Windows 11 Pro Version 10.0.22000.318

#### ★ Optional

![image](https://user-images.githubusercontent.com/5206262/141603968-9689623f-b4ee-4436-8797-857ce26bbb3a.png)

**Comment by Finii** [comment]
I do not have Pragmata Pro here, but I tried Pragmata Pro Bold Italic.
I see no issue there:
![image](https://user-images.githubusercontent.com/16012374/142841350-41969694-1102-4dcb-bab6-7623d018d75c.png)

My guess would be that you use PragmataPro (without NF) and the symbols are added by the system from another font, not matching in width. I'm not sure how the windows fonts are mapped (?) into the virtual Linux.
Do you see the issue when you use Writer or some other text processor directly on windows and type some text? Like I did for the screenshot above. The codes I used are uF74A and uF31B. Or copy and paste them from here:   and change the font.

**Comment by ToasterDEV** [comment]
Hey there!

> My guess would be that you use PragmataPro (without NF) and the symbols are added by the system from another font, not matching in width. I'm not sure how the windows fonts are mapped (?) into the virtual Linux.
Actually, the font is applied only from the Windows side, as the terminal emulator is packaged as an UWP app, so I just patched the original font using the cited command and saw the result regardless of if the console host was on the WSL VM or running under Windows.

Actually, the font is applied only from the Windows side, as the terminal emulator is packaged as an UWP app, so I just patched the original font using the cited command and saw the result regardless of if the console host was on the WSL VM or running under Windows.

Then again, when I'm using Caskaydia Cove NF the font does not show the same issue:
![image](https://user-images.githubusercontent.com/5206262/142957389-1a300601-5c1e-4111-829e-6cfc46b95edf.png)
Patched PragmataPro
![image](https://user-images.githubusercontent.com/5206262/142957702-21456424-3e2a-4545-8046-b36382dcd09b.png)
Patched Cascadia Code (Downloaded from site, not manually patched)
> Do you see the issue when you use Writer or some other text processor directly on windows and type some text? Like I did for the screenshot above. The codes I used are uF74A and uF31B. Or copy and paste them from here:   and change the font.

I just tried changing my font in Sublime Text 4 within Windows and aside from oddly a difference in aliasing, the icons are rendering correctly.

![image](https://user-images.githubusercontent.com/5206262/142957082-ab4b8d7d-d0e7-440e-9a70-16d1b79f3c11.png)
![image](https://user-images.githubusercontent.com/5206262/142957389-1a300601-5c1e-4111-829e-6cfc46b95edf.png)

With Nerd Font patch

Comparison:
![image](https://user-images.githubusercontent.com/5206262/142957183-2e0e6815-c8fa-49c9-b483-2d2717ccc87b.png)
Patched
![image](https://user-images.githubusercontent.com/5206262/142957212-da9a6b0e-bf1c-4add-acd3-7d3ac70c5cb1.png)
Unpatched


In this case, would it be an issue with the Terminal App itself, or is there something else I should be checking?

Thanks for the help!

**Comment by ToasterDEV** [comment]
Just found out some new stuff, perhaps something's broken with oh-my-posh, since the icons show up correctly when displaying some of the other themes.
![image](https://user-images.githubusercontent.com/5206262/142958339-d3682572-2b85-43f0-80a9-f55066491801.png)
![image](https://user-images.githubusercontent.com/5206262/142958353-ca5ad2ec-8929-41f3-bf15-3f30ae607a00.png)
![image](https://user-images.githubusercontent.com/5206262/142958371-721ee86b-1ea2-4bf4-a25e-cfa3fe6210d9.png)
![image](https://user-images.githubusercontent.com/5206262/142958385-4535e70b-53d4-44bd-b1aa-0363ea4acb2c.png)
![image](https://user-images.githubusercontent.com/5206262/142958398-67579c16-ec2c-4709-9520-3b82aa3d4ebc.png)
![image](https://user-images.githubusercontent.com/5206262/142958455-79b71221-ad83-455c-bac9-e6fef94ace60.png)
Given it seems the icons are rendering correctly, just not being shown fully, should I go over to the oh-my-posh devs and check with them whether there's an issue with the font or the theme?

**Comment by Finii** [comment]
Well, Cascadia does not include the sybols that we see here:
![image](https://user-images.githubusercontent.com/16012374/142982570-b709ba56-e070-4506-8b34-f8ba191401a6.png)

Checked `CascadiaMonoPL-Regular.otf` of Release 2110.31 from [MS's git repo](https://github.com/microsoft/cascadia-code/releases), the blue box would be `Ubuntu uF31B`:
![image](https://user-images.githubusercontent.com/16012374/142983120-5e4d0a41-174e-482c-8890-4ce6ccb770ab.png)

So there is some magic font substitution at work.

----
Given your image
![image](https://user-images.githubusercontent.com/16012374/142985551-2d29a330-d34c-40ec-b976-550448db42c7.png)
I believe you use the non-mono patched version.

I tried these two command lines
```
fontforge -script ../../font-patcher -l -w -c -s --careful PragmataPro\ Bold\ Italic.ttf
fontforge -script ../../font-patcher -w -c --careful PragmataPro\ Bold\ Italic.ttf
```
Which result in these two fonts:
* `PragmataPro Bold Italic Nerd Font Complete Mono Windows Compatible.ttf`
* `PragmataPro Bold Italic Nerd Font Complete Windows Compatible.ttf`

Examining the fonts shows this:
![image](https://user-images.githubusercontent.com/16012374/142986291-c44095ec-a1de-4054-a135-ee828a29706d.png)
The normal width seen with `X` is obviously 1060.
The left logo is from the non-mono font (and is too big, like in your problematic cases), whiule the right logo has been scaled down to 1060 in the patching process.

So my conclusion here is that either you use the non-mono-patched version ;-> or the source font you use contains some glyphs that confuse our 'how wide is a capital letter' algorithm so that the logo is scaled too big. Against that assumption is that the logo width exactly matches the unscaled logo width.

----

> whether there's an issue with the font or the theme?

I would first make sure the font is patched correctly. Use `writer` (or `word` or whatever) to write some text with the logo. The width of all chars MUST be the same for console use; use the `mono` patched font (as you said you did). If the icons are rendered wider than a capital `X` something is wrong with the patched font. (In your example above the two icons `folder` and `ubuntu` are as wide as four ordinary characters `Hell`)

Maybe check the rendered font with fontforge (Windows version [is available](https://fontforge.org/en-US/downloads/windows-dl/)). Open the patched font, open `view` `Goto` and search for `ubuntu`. Double click the small preview. Check the width on the ruler...

If you are sure that use use the `mono` patched font, and still the `ubuntu` logo is wider than `X`, the only viable solution I see is that someone has a look on your source font, why it confuses `font-patcher`. But that is of course problematic as it is not a free font :->

If fontforge on windows with your patched font shows that `ubuntu` is of the same width as `X`, but you problem persists in `writer` there is possibly something wrong with the fonts, try uninstalling all Pragmatas (patched and unpatched), reboot, and install just the patched version. That should work (cough).
If it looks ok in `writer` (i.e. is monospaced: `ubuntu` is same width as `X`) but not in the ZSH or what is it? Only then it might be a powerline / theme / whatever problem.

Generally I would try to stick to one system first and make that work, the virtual Linux adds a lot of complexity I guess (never used that). So maybe first try with PowerShell or something directly on windows. Look out for 'too wide' symbols, that is what breaks your shell prompt.

**Comment by ToasterDEV** [comment]
Hello there!



> Checked `CascadiaMonoPL-Regular.otf` of Release 2110.31 from [MS's git repo](https://github.com/microsoft/cascadia-code/releases), the blue box would be `Ubuntu uF31B`: ![image](https://user-images.githubusercontent.com/16012374/142983120-5e4d0a41-174e-482c-8890-4ce6ccb770ab.png)
> 
> So there is some magic font substitution at work.

The font I was referring to was downloaded from the following link:
https://github.com/ryanoasis/nerd-fonts/releases/download/v2.1.0/CascadiaCode.zip
Not the original one from Microsoft, but I see where the confusion could have come from, sorry.



> If you are sure that use use the `mono` patched font, and still the `ubuntu` logo is wider than `X`, the only viable solution I see is that someone has a look on your source font, why it confuses `font-patcher`. But that is of course problematic as it is not a free font :->

Given that I'm converting the whole family folder and manually choosing the monospaced font, would it be an option to send over a copy of one of the files both pre and post patching for analysis?

> the source font you use contains some glyphs that confuse our 'how wide is a capital letter' algorithm so that the logo is scaled too big. Against that assumption is that the logo width exactly matches the unscaled logo width.

Would this assumption still hold if I were to forego the --careful switch? I just tried to repatch anew without it and the same results appeared.

> If fontforge on windows with your patched font shows that `ubuntu` is of the same width as `X`, but you problem persists in `writer` there is possibly something wrong with the fonts, try uninstalling all Pragmatas (patched and unpatched), reboot, and install just the patched version. That should work (cough).
> If it looks ok in `writer` (i.e. is monospaced: `ubuntu` is same width as `X`) but not in the ZSH or what is it? Only then it might be a powerline / theme / whatever problem.

Okay, just tried to look through both `X` and ` ` through FontForge, both show an edge to edge width of 1024pxˀ (Not sure if pixels or points)
![image](https://user-images.githubusercontent.com/5206262/143669851-1816bb3b-a216-453f-a846-5c1cd133009a.png)
![image](https://user-images.githubusercontent.com/5206262/143669859-344a2207-fbc5-433d-bad9-a1ab70000b84.png)


The only thing that looked odd when checking on Sublime Text was that the Ubuntu glyph seemingly was two characters even though I specified single-width on the command (though perhaps this is a Unicode oddity instead of a bug with my setup)
![image](https://user-images.githubusercontent.com/5206262/143669073-9178b11b-1814-42b2-8ec8-9a9e35304670.png)


> Generally I would try to stick to one system first and make that work, the virtual Linux adds a lot of complexity I guess (never used that). So maybe first try with PowerShell or something directly on windows. Look out for 'too wide' symbols, that is what breaks your shell prompt.
The terminal emulator is the one handling the font in this case ([Windows Terminal](https://docs.microsoft.com/en-us/windows/terminal/customize-settings/profile-appearance#font-face)), but just to check I tried setting the same theme over with Windows PowerShell and the results are similar (though with a sliced Apple logo instead of the Windows logo for some reason)

![image](https://user-images.githubusercontent.com/5206262/143669129-5bfd709c-729f-44ee-a27c-00e216e188ac.png)
I'm assuming perhaps the same thing from the previous picture applies here, but honestly I'm not too sure if the problem lies within my powerline handler, given that Caskaydia Cove Nerd Font is able to render the glyphs without issue considering all else as equal.

At the moment my last resort would be to try and remove all variants of the font, then rebooting and installing only the patched versions as instructed, but I'll do that and report in the next comment to make sure I don't lose what I have already written.

Cheers!

**Comment by ToasterDEV** [comment]
Okay, returning from reboot and installing the fonts made with this command:
`docker run -v /mnt/c/Users/ToasterDEV/OneDrive/Documents/Backups/Typography/PragmataPro:/in -v /mnt/c/Users/ToasterDEV/OneDrive/Documents/Backups/Typography/PragmataProNerdFonts/out:/out nerdfonts/patcher -l -w -c -s --progressbars --careful`
It oddly seems to have fixed the icons! (albeit a bit smaller than expected, running without the single width switch leads to previous results.)
![image](https://user-images.githubusercontent.com/5206262/143671604-9d701508-b216-4024-9639-089c699b49bb.png)
![image](https://user-images.githubusercontent.com/5206262/143671630-e9f8e086-f7b2-4cbe-aa26-6275e523ba05.png)

I'm puzzled that having the other fonts installed could have led to issues, but at least I can see the glyphs as expected now without any major setbacks.

Given that (from my understanding) the issue was on Windows' side (OS or Terminal), I'd like to close the issue, though I'm not entirely sure why the proposed solution worked at all, since that the same doesn't happen between Cascadia Code and Caskaydia Cove NF when both are installed at the same time, so I'm open to testing if it's still required.

Otherwise, I guess I'll go ahead and report the bug to the Windows Terminal team, and in the meantime I'd like to thank you for the help!

**Comment by shogerr** [comment]
Setting `antialiasingMode` to either `grayscale` or `cleartype` in the Windows Terminal settings may change the glyph rendering for you.

If I set `antialiasingMode='aliased'` I will end up seeing cutoff glyphs in the powerline for `oh-my-posh`.

**Comment by Finii** [comment]
I guess this could be haven been fixed by `v2.2.2`.

**Comment by kierun** [comment]
Just for information, I have patched PragmataPro as well and this is what I have:

![screenshot](https://user-images.githubusercontent.com/3079575/212333368-19722f7e-5c58-42bf-89c9-da0120bdc979.png)


and

![screenshot-1](https://user-images.githubusercontent.com/3079575/212333139-25d455c7-118a-475e-9fa9-ff58d60597a2.png)

I do not see any issues… ¯\\\_(ツ)\_/¯

**Comment by Finii** [comment]
Thanks for checking. Did you use `Windows Terminal`?

The problem with these old-ish issues is that some might be because of the font and that has been fixed, but others might be caused by terminal program bugs, that were fixed now.

`v2.3.0` brings some scaling changes, I hope that solves more issues. ;-) (of course ;-D )

**Comment by kierun** [comment]
The font was patched using `#bbdb7377df` so a little forward of `v2.2.2`… I guess the fix could be before that.

I did not try on Windows as I mainly use Linux.

It was just a “well, it worked for me…” which can be less than useful. Although, it does show that under my case, it did work, so it's a more tricky bug to find. Hopefully, `v2.3.0` fixed it! :grinning:
